File: app/src/main/java/com/aatthhuurr/mobisuas_2411500025/PelaporanActivity.java
package com.aatthhuurr.mobisuas_2411500025;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.location.Location;
import android.os.Bundle;
import android.provider.MediaStore;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.AutoCompleteTextView;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.Toast;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.textfield.TextInputEditText;

public class PelaporanActivity extends AppCompatActivity {

    private ImageButton btnBack, btnAddPhoto;
    private TextInputEditText edtJudul, edtSearchMaps, edtKronologis;
    private AutoCompleteTextView dropKategori;
    private MaterialButton btnGetLocation, btnLaporkan;
    private ImageView imgBuktiPreview;
    private View cardMediaPreview; // Menambahkan variabel layout preview data

    private static final int CAMERA_REQUEST_CODE = 101;
    private static final int LOCATION_REQUEST_CODE = 102;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_pelaporan);

        // TODO 1: Casting ID XML
        btnBack = findViewById(R.id.btnBack);
        btnAddPhoto = findViewById(R.id.btnUploadMedia);
        edtJudul = findViewById(R.id.edtJudul);
        edtSearchMaps = findViewById(R.id.edtSearchMaps);
        edtKronologis = findViewById(R.id.edtKronologis);
        dropKategori = findViewById(R.id.dropKategori);
        btnGetLocation = findViewById(R.id.btnGetLocation);
        btnLaporkan = findViewById(R.id.btnLaporkan);
        imgBuktiPreview = findViewById(R.id.imgBuktiPreview);
        cardMediaPreview = findViewById(R.id.cardMediaPreview); // Pastikan ID ini ada di XML kalian gess

        // TODO 2: Membuat daftar Kategori Pengaduan di Dropdown
        String[] daftarKategori = {"Tambang Ilegal", "Polusi Lingkungan", "Kebakaran Hutan", "Fasilitas Rusak", "Lainnya"};
        ArrayAdapter<String> adapterKategori = new ArrayAdapter<>(this, android.R.layout.simple_dropdown_item_1line, daftarKategori);
        dropKategori.setAdapter(adapterKategori);

        // TODO 3: Fungsikan Ambil Foto dari Kamera
        btnAddPhoto.setOnClickListener(v -> {
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.CAMERA}, CAMERA_REQUEST_CODE);
            } else {
                bukaKamera();
            }
        });

        // TODO 4: Ambil Koordinat GPS Terkini
        btnGetLocation.setOnClickListener(v -> {
            if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, LOCATION_REQUEST_CODE);
            } else {
                ambilLokasiGps();
            }
        });

        // TODO 5: Tombol Kirim Laporan ke API
        btnLaporkan.setOnClickListener(v -> {
            String judul = edtJudul.getText().toString().trim();
            String kategori = dropKategori.getText().toString().trim();
            String kronologis = edtKronologis.getText().toString().trim();

            if (judul.isEmpty() || kategori.isEmpty() || kronologis.isEmpty()) {
                Toast.makeText(this, "Form laporan tidak boleh kosong!", Toast.LENGTH_SHORT).show();
            } else {
                // Di sini nanti proses tembak data ke server backend kalian
                Toast.makeText(this, "Laporan berhasil terkirim ke server!", Toast.LENGTH_SHORT).show();
                finish();
            }
        });

        // TODO 6: Tombol Kembali
        btnBack.setOnClickListener(v -> finish());
    }

    private void bukaKamera() {
        Intent cameraIntent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
        startActivityForResult(cameraIntent, CAMERA_REQUEST_CODE);
    }

    private void ambilLokasiGps() {
        // Menggunakan LocationManager bawaan Android agar tidak perlu library tambahan luar gess
        android.location.LocationManager locationManager = (android.location.LocationManager) getSystemService(LOCATION_SERVICE);
        if (locationManager != null) {
            if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
                Location lastKnownLocation = locationManager.getLastKnownLocation(android.location.LocationManager.GPS_PROVIDER);
                if (lastKnownLocation != null) {
                    double latitude = lastKnownLocation.getLatitude();
                    double longitude = lastKnownLocation.getLongitude();
                    // Set teks lokasi berupa koordinat ke kolom maps
                    edtSearchMaps.setText(latitude + ", " + longitude);
                    Toast.makeText(this, "Lokasi GPS berhasil didapatkan!", Toast.LENGTH_SHORT).show();
                } else {
                    Toast.makeText(this, "Nyalakan GPS HP kamu terlebih dahulu gess!", Toast.LENGTH_SHORT).show();
                }
            }
        }
    }

    // Menangkap hasil foto setelah kamera ditutup
    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == CAMERA_REQUEST_CODE && resultCode == RESULT_OK && data != null) {
            Bitmap imageBitmap = (Bitmap) data.getExtras().get("data");
            if (cardMediaPreview != null) {
                cardMediaPreview.setVisibility(View.VISIBLE); // Munculkan wadah frame preview
            }
            imgBuktiPreview.setImageBitmap(imageBitmap); // Pasang fotonya
        }
    }
}

File: app/src/main/java/com/aatthhuurr/mobisuas_2411500025/PelaporanActivity.kt
package com.aatthhuurr.mobisuas_2411500025

import android.net.Uri
import android.os.Bundle
import android.view.View
import android.widget.ArrayAdapter
import android.widget.AutoCompleteTextView
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.cardview.widget.CardView
import com.aatthhuurr.mobisuas_2411500025.api.RetrofitClient
import com.aatthhuurr.mobisuas_2411500025.helper.SharedPrefManager
import com.aatthhuurr.mobisuas_2411500025.model.LaporanResponse
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MarkerOptions
import com.google.android.material.button.MaterialButton
import com.google.android.material.textfield.TextInputEditText
import okhttp3.MediaType
import okhttp3.MultipartBody
import okhttp3.RequestBody
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class PelaporanActivity : AppCompatActivity(), OnMapReadyCallback {

    private var mGoogleMap: GoogleMap? = null
    private var uriGambarTerpilih: Uri? = null
    private lateinit var imgBuktiPreview: ImageView
    private lateinit var cardMediaPreview: CardView

    // Registrasi fungsi untuk menangkap hasil pemilihan gambar dari galeri
    private val bukaGaleriLauncher = registerForActivityResult(
        ActivityResultContracts.GetContent()
    ) { uri: Uri? ->
        if (uri != null) {
            uriGambarTerpilih = uri
            // Tampilkan komponen preview dan masukkan gambarnya
            cardMediaPreview.visibility = View.VISIBLE
            imgBuktiPreview.setImageURI(uri)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_pelaporan)

        val btnBack = findViewById<ImageButton>(R.id.btnBack)
        val edtJudul = findViewById<TextInputEditText>(R.id.edtJudul)
        val dropKategori = findViewById<AutoCompleteTextView>(R.id.dropKategori)
        val edtSearchMaps = findViewById<TextInputEditText>(R.id.edtSearchMaps)
        val btnGetLocation = findViewById<MaterialButton>(R.id.btnGetLocation)
        val edtKronologis = findViewById<TextInputEditText>(R.id.edtKronologis)
        val btnUploadMedia = findViewById<MaterialButton>(R.id.btnUploadMedia)
        val btnLaporkan = findViewById<MaterialButton>(R.id.btnLaporkan)

        // Inisialisasi komponen pratinjau foto dari XML asli kamu
        imgBuktiPreview = findViewById(R.id.imgBuktiPreview)
        cardMediaPreview = findViewById(R.id.cardMediaPreview)

        val mapFragment = supportFragmentManager.findFragmentById(R.id.mapFragment) as? SupportMapFragment
        mapFragment?.getMapAsync(this)

        val listNamaKategori = arrayOf("Tambang Ilegal", "Kebakaran", "Bencana", "Gangguan PDAM", "Penumpukan Sampah")
        val adapterKategori = ArrayAdapter(this, android.R.layout.simple_dropdown_item_1line, listNamaKategori)
        dropKategori.setAdapter(adapterKategori)

        btnBack.setOnClickListener { finish() }

        btnGetLocation.setOnClickListener {
            edtSearchMaps.setText("Pangkalpinang, Bangka Belitung")
            val lokasiBabel = LatLng(-2.1299, 106.1138)
            mGoogleMap?.clear()
            mGoogleMap?.addMarker(MarkerOptions().position(lokasiBabel).title("Lokasi Kejadian"))
            mGoogleMap?.animateCamera(CameraUpdateFactory.newLatLngZoom(lokasiBabel, 15f))
            Toast.makeText(this, "Lokasi berhasil disematkan", Toast.LENGTH_SHORT).show()
        }

        // AKSI TOMBOL UPLOAD FOTO: Membuka galeri sistem Android untuk memilih file gambar
        btnUploadMedia.setOnClickListener {
            bukaGaleriLauncher.launch("image/*")
        }

        btnLaporkan.setOnClickListener {
            val javaJudul = edtJudul.text.toString().trim()
            val javaLokasi = edtSearchMaps.text.toString().trim()
            val javaKronologis = edtKronologis.text.toString().trim()
            val kategoriTerpilih = dropKategori.text.toString()

            val prefManager = SharedPrefManager(this)
            val idUserTerlogin = prefManager.getIdUser()

            val idKategori = when (kategoriTerpilih) {
                "Tambang Ilegal" -> 1
                "Kebakaran" -> 2
                "Bencana" -> 3
                "Gangguan PDAM" -> 4
                "Penumpukan Sampah" -> 5
                else -> 0
            }

            if (javaJudul.isEmpty() || javaLokasi.isEmpty() || javaKronologis.isEmpty() || idKategori == 0) {
                Toast.makeText(this, "Semua data formulir pelaporan wajib diisi!", Toast.LENGTH_SHORT).show()
            } else if (idUserTerlogin == -1) {
                Toast.makeText(this, "Sesi habis, silakan lakukan login ulang", Toast.LENGTH_SHORT).show()
            } else {
                // DIPERBAIKI: Sekarang mengoper variabel uriGambarTerpilih ke fungsi pengirim
                kirimLaporanKeServer(idUserTerlogin, idKategori, javaJudul, javaLokasi, javaKronologis, uriGambarTerpilih)
            }
        }
    }

    override fun onMapReady(googleMap: GoogleMap) {
        mGoogleMap = googleMap
        val defaultBabel = LatLng(-2.1299, 106.1138)
        mGoogleMap?.moveCamera(CameraUpdateFactory.newLatLngZoom(defaultBabel, 12f))
    }

    // DIPERBAIKI: Fungsi kirim laporan berbasis Multipart Form-Data untuk mengunggah file biner
    private fun kirimLaporanKeServer(idUser: Int, idKategori: Int, judul: String, lokasi: String, kronologis: String, fileUri: Uri?) {

        // 1. Konversi data teks biasa menjadi RequestBody Multipart
        val rbIdUser = RequestBody.create(MediaType.parse("text/plain"), idUser.toString())
        val rbIdKategori = RequestBody.create(MediaType.parse("text/plain"), idKategori.toString())
        val rbJudul = RequestBody.create(MediaType.parse("text/plain"), judul)
        val rbLokasi = RequestBody.create(MediaType.parse("text/plain"), lokasi)
        val rbKronologis = RequestBody.create(MediaType.parse("text/plain"), kronologis)

        // 2. Olah file gambar fisik dari URI galeri menggunakan ContentResolver
        var bodyGambar: MultipartBody.Part? = null
        if (fileUri != null) {
            try {
                val inputStream = contentResolver.openInputStream(fileUri)
                val bytes = inputStream?.readBytes()
                inputStream?.close()

                if (bytes != null) {
                    val requestFile = RequestBody.create(MediaType.parse("image/*"), bytes)
                    // "bukti_foto" disamakan dengan input $_FILES['bukti_foto'] di PHP backend
                    bodyGambar = MultipartBody.Part.createFormData(
                        "bukti_foto",
                        "laporan_${System.currentTimeMillis()}.jpg",
                        requestFile
                    )
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        // 3. Eksekusi pengiriman data campuran ke server Apache via Retrofit
        RetrofitClient.instance.insertLaporan(rbIdUser, rbIdKategori, rbJudul, rbLokasi, rbKronologis, bodyGambar)
            .enqueue(object : Callback<LaporanResponse> {
                override fun onResponse(call: Call<LaporanResponse>, response: Response<LaporanResponse>) {
                    if (response.isSuccessful && response.body() != null) {
                        val res = response.body()!!
                        Toast.makeText(this@PelaporanActivity, res.message, Toast.LENGTH_SHORT).show()
                        if (res.status) {
                            finish() // Berhasil menyimpan, tutup halaman form
                        }
                    } else {
                        Toast.makeText(this@PelaporanActivity, "Gagal memproses respon server", Toast.LENGTH_SHORT).show()
                    }
                }

                override fun onFailure(call: Call<LaporanResponse>, t: Throwable) {
                    Toast.makeText(this@PelaporanActivity, "Koneksi gagal: ${t.message}", Toast.LENGTH_SHORT).show()
                }
            })
    }
}